### ISSUE TITLE
Sony a5000 stuck Connecting...

### ISSUE THREAD

**Issue by bytestream** [ThibaudM/timelapse-sony#29]
I used your app last year and had great success. I've downloaded it again on a new phone (Andriod 7) and noticed the UI has been updated. After starting the Smart Remote Control camera app on the Timelapse app on my phone the camera gets stuck in a Connecting... state.

Interestingly all three connection states in the timelapse app show a green tick and I can proceed to the next step. However... when clicking the "Take picture" button it says "We couldn't send or we didnt receive instructions to/from the camera".

Smart Remote Control camera app is running the latest version (4.30), last year I was running 4.10.

Any steps I can follow to debug?

**Comment by Gertsch** [comment]
I have got the same issue with my A6000....
Too bad it's not working.
Have tried it on three different android devices

**Comment by retracked** [comment]
I took some time to look into this. There seem to be missing some steps in the code to get a fully connected state with active live view and the possibility to take pictures.

I'm not completely sure what's missing, but in some rare cases when you close and reopen the app (maybe multiple times) it works. The correct initialization (for some cameras) happens only at startup and apparently not after the wifi connection. I guess @ThibaudM will have to look into this.

**Comment by Ekkaveli** [comment]
The steps which are missing are "startRecMode" and "setShootMode". Tried to debug the code and the method CameraApi.initialize() is called before there is a valid connection. If you call it after CameraApi.testConnection it works fine. If I have got time this week I can make a merge-request with a stable connection procedure (tested on alpha-6000).

**Comment by retracked** [comment]
A had a similar impression. But then again there were cases when it worked for me - but I think only while debugging, so the camera had more time to react in between the commands.

Did you have a look at [CameraAPI.initalize()](https://github.com/ThibaudM/timelapse-sony/blob/7dde7b40fe65055c8a71207683e0a60ea994bf95/app/src/main/java/com/thibaudperso/sonycamera/sdk/CameraAPI.java#L42)? There it would be in correct order. But if I remember correctly, the method didn't run through for me because `mIsDeviceInitialized==true` was fulfilled.

**Comment by tcristo** [comment]
I have the a5100.  For my camera, the two functions located in the CameraAPI.intialize() method must proceed the CameraAPI.startLiveView() method.  It appears initialize() is currently being called too early in the process.  When I remove the initialize() method from the setDevice() method and place it at the head of the startLiveView() method, it handles all use cases that I tried.  There is some cleanup that needs to be done to make sure the mIsDeviceInitialized = false; is set during closeout.  When I make the above changes, the app now work consistently for me.

**Comment by ThibaudM** [comment]
Sorry for the late answer,

In fact, I received a non-negligible amount of emails where people encountered problem only since the v3.
I could not reproduce this behavior with my rx100 mk2, so your answers are important for me.

@Ekkaveli and @tcristo, you are right, the `initialize` method was called too early. Now I call them only at the end of connection step. 

I committed the modifications in 99e38fb. They work for me but I need someone to test it. I've attached the apk file to the commit (`v3.0.3`). If the problem persists, you can use `v3.0.3-logger`

@retracked I removed this condition in v3, that was senseless. Now, these methods are called for each kind of device.

**Comment by Deph0** [comment]
v3.0.3 crashes when attempting to connect to the wifi of the device, startscreen and continues to do so in a loop if you attempt to restart the app.

**Comment by bytestream** [comment]
Yeah, same as @Deph0. I did manage to get it working occasionally but it's pretty hit and miss (using `a5000 (ILCE-5000)`).

Emailed you the log.

**Comment by ThibaudM** [comment]
Euhm, that does not seem to be good :/
Is someone can send me the logcat crash? (if possible, with the last version - cccb2de)

From the logs, I can see that the setShootMode is not taken into account: "Not Available Now".
In cccb2de (v3.0.4), I send the "setShootMode" command only when the "startRecMode" finished. 

```
23:32:46.473: sendRequest: {"version":"1.0","id":1,"method":"getVersions","params":[]}
23:32:46.544: result: {"result":[["1.0","1.1","1.2","1.3","1.4"]],"id":1}
23:33:05.761: sendRequest: {"version":"1.0","id":2,"method":"startRecMode","params":[]}
23:33:05.763: sendRequest: {"version":"1.0","id":3,"method":"setShootMode","params":["still"]}
23:33:05.857: sendRequest: {"version":"1.0","id":4,"method":"startLiveview","params":[]}
23:33:05.899: result: {"result":[0],"id":2}
23:33:05.993: result: {"id":3,"error":[1,"Not Available Now"]}
23:33:07.004: result: {"result":["http:\/\/192.168.122.1:8080\/liveview\/liveviewstream"],"id":4}
[...]
23:33:57.188: sendRequest: {"version":"1.0","id":63,"method":"actTakePicture","params":[]}
23:33:57.227: result: {"id":63,"error":[1,"Not Available Now"]}
23:33:57.356: sendRequest: {"version":"1.0","id":64,"method":"actTakePicture","params":[]}
23:33:57.395: result: {"id":64,"error":[1,"Not Available Now"]}
23:33:58.669: sendRequest: {"version":"1.0","id":65,"method":"actZoom","params":["out","1shot"]}
23:33:58.727: result: {"id":65,"error":[1,"Not Available Now"]}
```

**Comment by tcristo** [comment]
The API documentation says that it is possible for you to get a callback result without the camera actually finish processing the command.  If you poll on getEvent after each command sent, you will notice there are times where the response is "Not Ready" and yet the callback from the original command has already returned and yet you are prepared to send another command to the camera.

There are times where I have seen a delay up to a half second before the getEvent cameraStatus returns IDLE.  The timeframe doesn't seem to be relative to the command being sent as far as I can tell.  It just happens sometimes.

**Comment by ThibaudM** [comment]
That's true, but unfortunately, the `getEvent` method is not implemented on my rx100 mk2.
About the IDLE status, I've noticed the same behavior with my camera. Sometimes the IDLE state remains more than one minute

**Comment by tcristo** [comment]
Not sure what you can do if the getEvent API doesn't work for your camera.  It seems pretty integral for the a5000 to getting it to work correctly. 

What I do after I make a call to most of the camera APIs is "block" on a getEvent response that provides an IDLE status.  For my camera model, this has proven effective 100% of the time.  Below is an example of what I am using in one of the states in the the sm method.  I would send you the entire project but it is in a pretty messy state currently.  I am transitioning the state machine to a queue list so I can use it on bracketing and smoothing functions which require multiple, rapid shots with slight changes to the camera in between the shots. 

          CHECK_CAMERA_INITIALIZED{
            @Override
            public void process(final StateMachineConnection sm) {
                Log.d(LOG_TAG, " ----------- CHECK_CAMERA_INITIALIZED process invoked -----------");

                if(!sm.mCameraIsIdle) {
                    Log.d(LOG_TAG, " ----------- Waiting on IDLE getEventResponse -----------");
                    sm.waitOngetEventResponse(sm);
                    return;
                }

                sm.mCameraInit.CheckCameraStatus(new CameraInit.ApiStatusListener() {
                    @Override
                    public void isConnected(boolean isConnected, boolean isIdle, Object configInfo) {
                        if (isIdle) {
                            sm.setCurrentState(State.CHECK_API_VERSION);
                            sm.mCameraInit.CheckCameraStatus(new CameraInit.ApiStatusListener() {
                                @Override
                                public void isConnected(boolean isConnected, boolean isIdle, Object configInfo) {

                                    if (isIdle) {
                                        sm.mCameraIsIdle = true;
                                    } else {
                                        sm.mCameraIsIdle = false;
                                    }
                                }
                            });
                        } else {
                            sm.mCameraIsIdle = false;
                            sm.setCurrentState(CHECK_CAMERA_INITIALIZED);
                        }
                    }
                });
            }
        },

The blocking method is nothing more than a delay followed by recalling the same state function.  I currently have it set to a 1 second delay between calls but that is arbitrary.  Setting it to 250 milliseconds is probably more optimal.

    private void waitOngetEventResponse(final StateMachineConnection sm) {

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                sm.mStateRegistry.apiAttempts++;
                sm.setCurrentState(getCurrentState());
               // mCurrentState.process(sm);
            }
        }, 1000);
    }

**Comment by alanbirtles** [comment]
3.0.4 works with the HX-90V from a Sony Xperia XZ, the play store version doesn't

**Comment by retracked** [comment]
I can confirm 3.0.4 basically works with NEX-6 and A7II. But there are some of cases when the app ends up in some strange state, e.g. when I exit the app using the home button. At this point i had to remove the app from memory and start it again.

Also, when two Sony Camera WIFI are in reach, in some cases there happens an app crash.

But in general 3.0.4 helps a lot since with the old version I couldn't even connect.

**Comment by ThibaudM** [comment]
Ok, that's a good news.
In fact, 7 days ago, I released the 3.0.4 version on the Google Play.
Next time, I will have a look at the strange states.

@bytestream Can I close this issue?

**Comment by bytestream** [comment]
@ThibaudM yeah, thanks. 3.0.4 is working on a5000.
